### PR TITLE
Capture command line in environment on Linux and Windows

### PR DESCRIPTION
**Pull request by keithc-ca**
Partially addresses #947.

### CODE REVIEW

File: runtime/j9vm/jvm.c [hunk 1]
```diff
@@ -452,6 +456,74 @@ static const J9SignalMapping signalMap[] = {
 	{NULL, J9_SIG_ERR}
 };
 
+static void
+captureCommandLine(void)
+{
+#define ENV_VAR_NAME "IBM_JAVA_COMMAND_LINE"
```
Comment (pshipton) [inline runtime/j9vm/jvm.c:462]: I suggest we call this OPENJ9_JAVA_COMMAND_LINE. It means changing the places that use it, but using OPENJ9 is more appropriate.

File: runtime/j9vm/jvm.c [hunk 2]
```diff
@@ -452,6 +456,74 @@ static const J9SignalMapping signalMap[] = {
...
+				break;
+			}
+			if ('\0' == buffer[length]) {
+				buffer[length] = ' ';
+			}
+		}
+		/* it's not fatal if setenv() fails, so don't bother checking */
+		setenv(ENV_VAR_NAME, buffer, 0 /* don't overwrite */);
+		free(buffer);
+done:
```
Comment (pshipton) [inline runtime/j9vm/jvm.c:504]: There is a memory leak error for some of the error cases that goto done.

File: runtime/j9vm/jvm.c [hunk 3]
```diff
@@ -452,6 +456,74 @@ static const J9SignalMapping signalMap[] = {
...
+		for (length -= 2;; length -= 1) {
+			if (0 == length) {
+				break;
+			}
+			if ('\0' == buffer[length]) {
+				buffer[length] = ' ';
+			}
+		}
+		/* it's not fatal if setenv() fails, so don't bother checking */
+		setenv(ENV_VAR_NAME, buffer, 0 /* don't overwrite */);
```
Comment (pshipton) [inline runtime/j9vm/jvm.c:502]: Why not overwrite to get the current command line? If you are running java from java, the inner java should set the environment variable appropriately.

File: runtime/j9vm/jvm.c [hunk 4]
```diff
@@ -452,6 +456,74 @@ static const J9SignalMapping signalMap[] = {
...
+	if (0 != getenv_s(&existingLength, NULL, 0, ENV_VAR_NAME)) {
+		/* getenv_s() failed */
+	} else if (0 != existingLength) {
+		/* variable is already defined */
+	} else {
+		const char *commandLine = GetCommandLineA();
+
+		if (NULL != commandLine) {
+			/* it's not fatal if _putenv_s() fails, so don't bother checking */
+			_putenv_s(ENV_VAR_NAME, commandLine);
```
Comment (pshipton) [inline runtime/j9vm/jvm.c:519]: Is there a wide version that can be used instead, along with GetCommandLineW()?

File: runtime/j9vm/jvm.c (hunk 1 above)
Comment (keithc-ca) [reply runtime/j9vm/jvm.c:462]: Changing the name introduces the potential that both are defined: DDR should continue to recognize `IBM_JAVA_COMMAND_LINE` that may be found in existing system dump files. I will adjust things so `OPENJ9_JAVA_COMMAND_LINE` takes precedence.

File: runtime/j9vm/jvm.c (hunk 2 above)
Comment (keithc-ca) [reply runtime/j9vm/jvm.c:504]: Good catch.

File: runtime/j9vm/jvm.c (hunk 3 above)
Comment (keithc-ca) [reply runtime/j9vm/jvm.c:502]: I hadn't thought of that use-case, but that makes sense: I'll change it to overwrite any existing value.

File: runtime/j9vm/jvm.c (hunk 4 above)
Comment (keithc-ca) [reply runtime/j9vm/jvm.c:519]: I'm not sure anything else is prepared for wide characters, but I'll investigate.

Comment (pshipton) [reply runtime/j9vm/jvm.c:519]: Using wide ensures we preserve all characters correctly in the environment variable. It doesn't affect fetching the environment variable. Any place that fetches the value can continue to work as it does now. Windows does any necessary conversion between wide and the format requested.

**Comment by pshipton** [comment]
I assume this does capture the expected command line? Seems to me the launcher forks a new process before calling the VM.

**Comment by keithc-ca** [comment]
> I assume this does capture the expected command line? Seems to me the launcher forks a new process before calling the VM.

I don't understand this comment.

On Linux, a process forks before calling a function in the `exec()` family. The command-line captured reflects the values provided to `exec()`.

On Windows, a function in the `CreateProcess()` family is used to create a new process. The value captured corresponds to the string provided to that function.

Things work as expected on both platforms.

File: runtime/j9vm/jvm.c (hunk 1 above)
Comment (keithc-ca) [reply runtime/j9vm/jvm.c]: Fixed in 229e47cda1681285905983db3863a8c897ba9eab.

File: runtime/j9vm/jvm.c (hunk 2 above)
Comment (keithc-ca) [reply runtime/j9vm/jvm.c:507]: Fixed in 229e47cda1681285905983db3863a8c897ba9eab.

File: runtime/j9vm/jvm.c (hunk 3 above)
Comment (keithc-ca) [reply runtime/j9vm/jvm.c]: Fixed in 229e47cda1681285905983db3863a8c897ba9eab.

File: runtime/j9vm/jvm.c (hunk 4 above)
Comment (keithc-ca) [reply runtime/j9vm/jvm.c]: Fixed in 229e47cda1681285905983db3863a8c897ba9eab.

**Comment by pshipton** [comment]
jenkins compile zlinux,win,win32 jdk8

**Review by pshipton — APPROVED** [review]

**Comment by pshipton** [comment]
Assuming no problems show up, we can cherry pick this for the 0.31 branch.

**Comment by keithc-ca** [comment]
> cherry pick this for the 0.31 branch

See #14525.

**Comment by pshipton** [comment]
@keithc-ca this should have a doc issue and doc updated. Can you pls create at least the issue asap. https://eclipse-openj9.github.io/openj9-docs/env_var/ needs to be updated.

**Comment by keithc-ca** [comment]
I created https://github.com/eclipse-openj9/openj9-docs/issues/925.

**Pull request merged**